### PR TITLE
image: avoid panic on nil fields when converting

### PR DESCRIPTION
**Pull request by q384566678**
When I execute the create command, there was an error, but there was no error when making the test. So need to improve the test content, and fix the error.

Signed-off-by: zhouhao <zhouhao@cn.fujitsu.com>

### CODE REVIEW

**Comment by q384566678** [comment]
@opencontainers/image-tools-maintainers PTAL

**Comment by q384566678** [comment]
@xiekeyang @coolljt0725 @cyphar @vbatts  PTAL

**Comment by cyphar** [comment]
What was the error? Can you include it in the commit message that fixes said error?

**Comment by q384566678** [comment]
The following error occurred:
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x48ef43]
goroutine 19 [running]:
panic(0x808060, 0xc420010130)
	/home/zhouhao/go/src/runtime/panic.go:500 +0x1ae
testing.tRunner.func1(0xc4200932c0)
	/home/zhouhao/go/src/testing/testing.go:579 +0x474
panic(0x808060, 0xc420010130)
	/home/zhouhao/go/src/runtime/panic.go:458 +0x271
github.com/opencontainers/image-tools/image.(*config).runtimeSpec(0xc420676240, 0x85e3f6, 0x6, 0xc4201d3920, 0x1a, 0x0)
	github.com/opencontainers/image-tools/image/_test/_obj_test/config.go:94 +0x1b3
github.com/opencontainers/image-tools/image.createBundle(0xa2f7e0, 0xc4204fc090, 0xc420543b20, 0xc4200df2a0, 0x13, 0x85e3f6, 0x6, 0x0, 0x0)
	github.com/opencontainers/image-tools/image/_test/_obj_test/image.go:414 +0x318
github.com/opencontainers/image-tools/image.createRuntimeBundle(0xa2f7e0, 0xc4204fc090, 0xc4200df2a0, 0x13, 0x85e384, 0x6, 0x85e3f6, 0x6, 0x0, 0x0, ...)
	github.com/opencontainers/image-tools/image/_test/_obj_test/image.go:348 +0x9ce
github.com/opencontainers/image-tools/image.CreateRuntimeBundleLayout(0xc4200df180, 0x16, 0xc4200df2a0, 0x13, 0x85e384, 0x6, 0x85e3f6, 0x6, 0x0, 0x0, ...)

Because there is no non-empty check on the optional fields in config.go.

File: image/config.go [hunk 1]
```diff
@@ -73,37 +73,41 @@ func (c *config) runtimeSpec(rootfs string) (*specs.Spec, error) {
...
 		}
+		s.Process.Env = append(s.Process.Env, c.Config.Env...)
+		s.Process.Args = append(s.Process.Args, c.Config.Entrypoint...)
+		s.Process.Args = append(s.Process.Args, c.Config.Cmd...)
 
-		gid, err := strconv.Atoi(ug[1])
-		if err != nil {
-			return nil, errors.New("config.User: unsupported gid format")
+		if len(s.Process.Args) == 0 {
+			s.Process.Args = append(s.Process.Args, "sh")
```
Comment (stevvooe) [inline image/config.go]: Do we make the assumption that `sh` is always available in a container?

**Comment by stevvooe** [comment]
A good title for this PR and the commit might be "image: avoid panic on nil fields when converting".

Also, please take the time to squash this into a single commit.

File: image/config.go (hunk 1 above)
Comment (q384566678) [inline image/config.go]: Yes, `Args` has at least one value.

**Comment by q384566678** [comment]
updated, PTAL.

File: image/config.go (hunk 1 above)
Comment (stevvooe) [inline image/config.go]: That is not what I asked.

Not all containers are going to have a `sh` binary present. Adding this, when it might not even exist, will certainly result in failure. This is also never mentioned in the specification. In the case that no args are provided, it should just not be set.

Comment (stevvooe) [inline image/config.go]: Fixing this is outside the scope of this PR. I've filed https://github.com/opencontainers/image-tools/issues/166.

Comment (q384566678) [inline image/config.go]: @stevvooe If there are no other questions, can we merge this patch first and then discuss the `sh` problem? Because if you don't merge this, the `create` function won't work.

File: image/config.go [hunk 2]
```diff
@@ -73,39 +73,43 @@ func (c *config) runtimeSpec(rootfs string) (*specs.Spec, error) {
 	var s specs.Spec
 	s.Version = specs.Version
 	// we should at least apply the default spec, otherwise this is totally useless
-	s.Process.Terminal = true
-	s.Root.Path = rootfs
-	s.Process.Cwd = "/"
-	if c.Config.WorkingDir != "" {
-		s.Process.Cwd = c.Config.WorkingDir
+	if s.Root != nil {
```
Comment (Mashimiao) [inline image/config.go]: I think this is not a right fix. Default Root is nil in Spec, we should initialize it as s.Root= &specs.Root{},  also same to Process.

Comment (q384566678) [inline image/config.go]: updated, thanks.

**Comment by stevvooe** [comment]
LGTM

**Comment by coolljt0725** [comment]
LGTM
